tests/python_package_test/test_sklearn.py

```diff
@@ -227,6 +227,21 @@ def test_joblib(self):
         pred_pickle = gbm_pickle.predict(X_test)
         np.testing.assert_allclose(pred_origin, pred_pickle)
 
+    def test_random_state_object(self):
+        X, y = load_iris(True)
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        state = np.random.RandomState(123)
+        clf = lgb.LGBMClassifier(n_estimators=10, subsample=0.5, subsample_freq=1, random_state=state)
+        self.assertIs(clf.random_state, state)
+        clf.fit(X_train, y_train)
+        y_pred1 = clf.predict(X_test, raw_score=True)
+        clf.fit(X_train, y_train)
+        y_pred2 = clf.predict(X_test, raw_score=True)
+        self.assertIs(clf.random_state, state)
+        self.assertRaises(AssertionError,
+                          np.testing.assert_allclose,
+                          y_pred1, y_pred2)
```

Comment:
two models that cleanly separate the test cases in `iris` could still be different, so I don't think this is a sufficient test that we're controlling randomness.

could you add a test that more thoroughly tests that we're getting the same exact model, like checking the feature importances? (https://github.com/microsoft/LightGBM/pull/2904/files#diff-1613c46333fcf60f940b13627ac833c7R218)

Reply:
I added new checks for equality of feature importance and booster state.
For the booster state, I now compare the output of the model_to_string() method.
Does that account for everything?

Reply:
@jameslamb Generally, I agree with you! But after merging #2804, two models with identical parameters should produce exactly the same results on the same machine even with multithreaded bagging.
> two models that cleanly separate the test cases in `iris` could still be different,

To prevent ideal class predictions, I set
-  small boosting rounds (to save CI time as well 😄 ),
-  `raw_score=True`, so we are not predicting classes.

Reply:
ohhhhhh I didn't notice `raw_score=True`. I thought scikit-learn interfaces for classification were always

* `predict()` = class labels
* `predict_proba()` = predict probability for each possible label

Then ok! I like the changes you pushed @lpfann. I don't have any requested  changes to the implementation, it looks solid esp. since it is backwards-compatible.

So from my perspective, this PR is good to go whenever you resolve https://github.com/microsoft/LightGBM/pull/2904#issuecomment-600795398 with @StrikerRUS 

Thanks again for the contribution!